pkg/ociwclayer/export.go

```diff
@@ -76,6 +78,27 @@ func writeTarFromLayer(ctx context.Context, r wclayer.LayerReader, w io.Writer)
 				return err
 			}
 		} else {
+			numberOfLinks, fileIDInfo, err := r.LinkInfo()
+			if err != nil {
+				return err
+			}
+			if numberOfLinks > 1 {
+				if linkName, ok := linkRecords[fileIDInfo.FileID]; ok {
+					// We've seen this file before, by another name, so put a hardlink in the tar stream.
+					hdr := backuptar.BasicInfoHeader(name, 0, fileInfo)
+					hdr.Mode = 0644
+					hdr.Typeflag = tar.TypeLink
+					hdr.Linkname = linkName
+					if err := t.WriteHeader(hdr); err != nil {
+						return err
+					}
+					continue
+				}
+
+				// All subsequent names for this file will be hard-linked to this name
+				linkRecords[fileIDInfo.FileID] = filepath.ToSlash(name)
+			}
+
 			err = backuptar.WriteTarFileFromBackupStream(t, r, name, size, fileInfo)
```

Comment:
I am a bit surprised to see that this is working. I am probably forgetting some details of layer export but in case of non-base layers we call `ExportLayer` which writes the layer contents into a backup stream format and that's why we use this `WriteTarFileFromBackupStream` here. But in case of base layers we are directly reading the layer files. We shouldn't need any backup stream readers at all? I also don't understand the need of `backupReader` in the `baseLayerReader`?

Reply:
Maybe I'm misremembering, but even for base layers, the actual file data in the tarball should be in Windows Backup format, I assume. So we use `backupReader` to pull the data from the on-disk file into the Windows Backup format, i.e. including ACLs/ADS/etc, so that streaming it back out restores all those things that tarballs don't carry.

I recall spending a while trying to work out how Windows Backup format stores hardlinks, before discovering in the OCI spec explicitly that hardlinks are to be handled at the tar layer, not the Windows Backup layer. Or something like that anyway. So I must have been under the impression that the data blobs in the tarball were in Windows Backup format to have bothered doing that research.

So short answer, `baseLayerReader.backupReader` is the inverse of `baseLayerWriter.bw`.

Reply:
Thanks for the detailed explanation @TBBle! You are right, the OCI layer tarball for a windows image layer seems to contain the backup formatted data. When reading an on-disk layer we first call `ExportLayer` on it which exports the contents of the `Files` directory into the backup format but keeps the contents of the other directories (`Hives`) in the standard file format. [This is why we check if the file is under the "Files" directory or not and either read it directly or read it via a backupReader](https://github.com/microsoft/hcsshim/blob/main/internal/wclayer/legacy.go#L243). The `WriteTarFileFromBackupStream` then just writes this data (all the layer contents, whether under "Files" or not are now in the backup format) into the tar with proper headers etc. 
When it comes to the base layer we don't call `ExportLayer` (it is not supported anyway), so we can just use a backup reader to read all of the layer files and pass that as it is to the `WriteTarFileFromBackupStream`. We probably should be able to rewrite the following code
```
func (r *baseLayerReader) Read(b []byte) (int, error) {
	if r.backupReader == nil {
		if r.currentFile == nil {
			return 0, io.EOF
		}
		return r.currentFile.Read(b)
	}
	return r.backupReader.Read(b)
}
```
as
```
func (r *baseLayerReader) Read(b []byte) (int, error) {
	if r.backupReader == nil {
		return 0, io.EOF
	}
	return r.backupReader.Read(b)
}
```
 since we never need to directly read from a file handle in case of base layer.

Reply:
Hi @ambarve!  Changes have been made. Please take another look.